File: 238-Product-of-Array-Except-Self/solution.java
public class Solution {
    public int[] productExceptSelf(int[] nums) {
        // Numbers:     2    3    4     5
        // Lefts:       1    2  2*3 2*3*4
        // Rights:  3*4*5  4*5    5     1
        int n = nums.length;
        int[] res = new int[n];
        res[0] = 1;
        for(int i=1; i<n; i++){
            res[i] = nums[i-1]*res[i-1];
        }
        int right = 1;
        for(int i=n-1; i>=0; i--){
            res[i] *= right;
            right *= nums[i];
        }
        return res;
    }
}

File: 238-Product-of-Array-Except-Self/solution.py
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        res = [1]*len(nums)
        for i in xrange(1, len(res)):
            res[i] = (res[i-1] * nums[i-1])
        right = 1
        for j in xrange(len(nums)-1, -1, -1):
            res[j] *= right
            right *= nums[j]
        return res
            
